agent/pkg/tunnel/tunnel.go

```diff
@@ -0,0 +1,66 @@
+package tunnel
+
+import (
+	"context"
+	"time"
+
+	"github.com/kubeedge/edgemesh/agent/pkg/tunnel/controller"
+	"github.com/kubeedge/edgemesh/common/constants"
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/peer"
+	ma "github.com/multiformats/go-multiaddr"
+	"k8s.io/klog/v2"
+)
+
+func (t *TunnelAgent) Run() {
+	for {
+		relay, err := controller.APIConn.GetPeerAddrInfo(constants.SERVER_ADDR_NAME)
+		if err != nil {
+			klog.Errorln("Failed to get tunnel server addr")
+			time.Sleep(5 * time.Second)
+			continue
+		}
+
+		if len(t.Host.Network().ConnsToPeer(relay.ID)) == 0 {
+			klog.Warningf("Connection between agent and server %v is not established, try connect", relay.Addrs)
+			retryTime := 0
+			for retryTime < 3 {
+				klog.Infof("Tunnel agent connecting to tunnel server")
+				err = t.Host.Connect(context.Background(), *relay)
+				if err != nil {
+					klog.Warningf("Connect to server err: %v", err)
+					time.Sleep(2 * time.Second)
+					retryTime++
+					continue
+				}
+
+				err = controller.APIConn.SetPeerAddrInfo(t.Config.NodeName, InfoFromHostAndRelay(t.Host, relay))
+				if err != nil {
+					klog.Warningf("Set peer addr info to secret err: %v", err)
+				}
```

Comment:
If setPeerAddInfo fail, do we need to retry?

Reply:
ok, i will add retry logic